source/hp/dof_handler.cc

```diff
@@ -1745,55 +1724,169 @@ namespace hp
-            // set in pre-refinement action)
-            //
-            // Note: Although one level is added to the DoFHandler
-            // levels, when the triangulation got one, for the buffer
-            // has_children this new level is not required, because
-            // the cells on the finest level never have
-            // children. Hence cell->has_children () will always
-            // return false on that level, which would cause shortcut
-            // evaluation of the following expression. Thus an index
-            // error in has_children should never occur.
-            if (cell->has_children() &&
-                !(*has_children[cell->level()])[cell->index()])
+            if (cell->refine_flag_set())
               {
-                // Set active_fe_index in children to the same value
-                // as in the parent cell. we can't access the
-                // active_fe_index in the parent cell any more through
-                // cell->active_fe_index() since that function is not
-                // allowed for inactive cells, but we can access this
-                // information from the DoFLevels directly
-                //
-                // we don't have to set the active_fe_index for ghost
-                // cells -- these will be exchanged automatically
-                // upon distribute_dofs()
-                for (unsigned int i = 0; i < cell->n_children(); ++i)
-                  if (cell->child(i)->is_locally_owned())
-                    cell->child(i)->set_active_fe_index(
-                      levels[cell->level()]->active_fe_index(cell->index()));
+                // Store the active_fe_index of each cell that will be refined
+                // to and distribute it later on its children.
+                refined_cells_fe_index.push_back(
+                  {cell->id(), cell->active_fe_index()});
+              }
+            else if (cell->coarsen_flag_set())
+              {
+                // From all cells that will be coarsened, determine their parent
+                // and calculate its proper active_fe_index, so that it can be
+                // set after refinement.
+                const auto &parent = cell->parent();
+                // Check if the active_fe_index for the current cell has been
+                // determined already.
+                if (std::find_if(
+                      coarsened_cells_fe_index.cbegin(),
+                      coarsened_cells_fe_index.cend(),
+                      [&parent](
+                        const std::pair<CellId, unsigned int> &element) {
+                        return element.first == parent->id();
+                      }) == coarsened_cells_fe_index.cend())
+                  {
+                    std::set<unsigned int> fe_indices_children;
+                    for (unsigned int child_index = 0;
+                         child_index < GeometryInfo<dim>::max_children_per_cell;
+                         ++child_index)
```

Comment:
make this also work for anisotropic refinement by using `child_index < parent->n_children()`